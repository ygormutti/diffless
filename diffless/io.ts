import * as fs from 'fs';
import { partition } from 'lodash';

import { charactersDiff } from './array-diff';
import { dynamicProgrammingLCS } from './lcs';
import { Change, ChangeLevel, ChangeType, Character, Document, Location, Position } from './model';
import { stripMargin } from './util';

class ChangeIndex {
    private index: Change[][][];

    constructor() {
        this.index = [];
    }

    add(location: Location, change: Change) {
        const { start, end } = location.range;
        this.addToPosition(start, change);
        this.addToPosition(end, change);
    }

    private addToPosition(position: Position, change: Change) {
        const { line, character } = position;
        if (!this.index[line]) { this.index[line] = []; }

        let changes = this.index[line][character];
        if (!changes) {
            changes = [];
            this.index[line][character] = changes;
        }

        changes.push(change);

        // sort by level
        changes.sort((a, b) => {
            return ((a.level * 100) - (b.level * 100)) + (a.type - b.type);
        });
    }

    get(position: Position): Change[] {
        const { line, character } = position;
        if (!this.index[line] || !this.index[line][character]) { return []; }
        return this.index[line][character];
    }
}

function buildAnnotatedDocumentHTML(document: Document, changeIndex: ChangeIndex) {
    const { characters, uri } = document;
    const pendingChanges = new Set();
    let html = `<pre class="${uri.replace(':', '_')}"><code>`;

    for (const character of characters) {
        html += buildPositionTags(character.position, changeIndex, pendingChanges);
        if (character.value === '\n') {
            html += '<br/>\n';
        } else {
            html += character.value;
        }
    }
    html += '</code></pre>';
    return html;
}

function buildPositionTags(
    position: Position,
    changeIndex: ChangeIndex,
    pendingChanges: Set<any>,
) {
    const changesHere = changeIndex.get(position);
    const [ending, starting] = partition(changesHere, c => pendingChanges.has(c));
    let html = '';
    for (const change of ending) {
        pendingChanges.delete(change);
        html += '</span>';
    }
    for (const change of starting) {
        pendingChanges.add(change);
        html += `<span class="${ChangeLevel[change.level]} ${ChangeType[change.type]}">`;
    }
    return html;
}

export function buildAnnotatedHTML(
    leftDocument: Document,
    rightDocument: Document,
    changes: Change[],
) {
    const leftChangeIndex = new ChangeIndex();
    const rightChangeIndex = new ChangeIndex();
    changes.forEach(change => {
        const { left, right } = change;
        if (left) { leftChangeIndex.add(left, change); }
        if (right) { rightChangeIndex.add(right, change); }
    });

    const leftPreCode = buildAnnotatedDocumentHTML(leftDocument, leftChangeIndex);
    const rightPreCode = buildAnnotatedDocumentHTML(rightDocument, rightChangeIndex);

    return stripMargin
        `<html>
    |<body>
    |<style>
    |    body {
    |        background-color: black;
    |        color: white;
    |    }
    |
    |    .Add {
    |        background-color: rgba(0,255,0,0.3);
    |    }
    |
    |    .Delete {
    |        background-color: rgba(255,0,0,0.3);
    |    }
    |
    |    .Move {
    |        background-color: rgba(0,0,255,0.3);
    |    }
    |
    |    .string_left {
    |        width: 50%;
    |        float: left;
    |    }
    |</style>
    |${leftPreCode}
    |${rightPreCode}
    |</body>
    |</html>`;
}

export function annotateWithChangesFile(leftPath: string, rightPath: string, changesPath: string) {
    const left = new Document('string:left', readTextFile(leftPath));
    const right = new Document('string:right', readTextFile(rightPath));

    const jsonChanges = JSON.parse(readTextFile(changesPath)) as JsonChange[];
    const changes = jsonChanges.map(c => toChange(c));

    return buildAnnotatedHTML(left, right, changes);
}

function readTextFile(path: string, encoding: string = 'utf8') {
    return fs.readFileSync(path, { encoding });
}

interface JsonChange {
    right?: Location;
    left?: Location;
    level: keyof typeof ChangeLevel;
    type: keyof typeof ChangeType;
}

function toChange(objFromJson: JsonChange): Change {
    return {
        ...objFromJson,
        level: ChangeLevel[objFromJson.level],
        type: ChangeType[objFromJson.type],
    };
}

export function annotateWithDiff(leftPath: string, rightPath: string) {
    const left = new Document('string:left', readTextFile(leftPath));
    const right = new Document('string:right', readTextFile(rightPath));

    const changes = charactersDiff(
        dynamicProgrammingLCS,
        left,
        right,
    );

    return buildAnnotatedHTML(left, right, changes);
}
