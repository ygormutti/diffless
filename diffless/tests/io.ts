import { annotateWithChangesFile, annotateWithDiff } from '../io';

describe('annotateWithDiff', () => {
    it('should work with diff', () => {
        const result = annotateWithDiff(
            'examples/json/before.json',
            'examples/json/single_map_reorder/after.json',
        );
        expect(result).toMatchSnapshot();
    });
});

describe('annotateWithChangesFile', () => {
    it('should match snapshot', () => {
        const result = annotateWithChangesFile(
            'examples/json/before.json',
            'examples/json/single_map_reorder/after.json',
            'examples/json/single_map_reorder/changes.json',
        );
        expect(result).toMatchSnapshot();
    });
});
