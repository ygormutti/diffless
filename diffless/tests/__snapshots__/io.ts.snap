// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`annotateWithChangesFile should match snapshot 1`] = `
"<html>
<body>
<style>
    body {
        background-color: black;
        color: white;
    }

    .Add {
        background-color: rgba(0,255,0,0.3);
    }

    .Delete {
        background-color: rgba(255,0,0,0.3);
    }

    .Move {
        background-color: rgba(0,0,255,0.3);
    }

    .string_left {
        width: 50%;
        float: left;
    }
</style>
<pre class=\\"string_left\\"><code>{<br/>
<span class=\\"Textual Delete\\">  </span><span class=\\"Textual Move\\">  <span class=\\"Lexical Move\\">\\"message\\": \\"Validation Failed\\"</span></span><span class=\\"Textual Delete\\"><span class=\\"Lexical Delete\\">,</span><br/>
  </span>  \\"errors\\": [<br/>
<span class=\\"Textual Delete\\">    </span>    {<br/>
<span class=\\"Textual Delete\\">      </span>      \\"resource\\": \\"Issue\\",<br/>
<span class=\\"Textual Delete\\">      </span>      \\"field\\": \\"title\\",<br/>
<span class=\\"Textual Delete\\">      </span>      \\"code\\": \\"missing_field\\"<br/>
<span class=\\"Textual Delete\\">    </span>    },<br/>
<span class=\\"Textual Delete\\">    </span>    {<br/>
<span class=\\"Textual Delete\\">      </span>      \\"resource\\": \\"Issue\\",<br/>
<span class=\\"Textual Delete\\">      </span>      \\"field\\": \\"milestone\\",<br/>
<span class=\\"Textual Delete\\">      </span>      \\"code\\": \\"invalid\\"<br/>
<span class=\\"Textual Delete\\">    </span>    }<br/>
<span class=\\"Textual Delete\\">  </span>  ]<br/>
}<br/>
</code></pre>
<pre class=\\"string_right\\"><code>{<br/>
  \\"errors\\": [<br/>
    {<br/>
      \\"resource\\": \\"Issue\\",<br/>
      \\"field\\": \\"title\\",<br/>
      \\"code\\": \\"missing_field\\"<br/>
    },<br/>
    {<br/>
      \\"resource\\": \\"Issue\\",<br/>
      \\"field\\": \\"milestone\\",<br/>
      \\"code\\": \\"invalid\\"<br/>
    }<br/>
  ]<span class=\\"Textual Add\\"><span class=\\"Lexical Add\\">,</span></span><br/>
<span class=\\"Textual Move\\">  <span class=\\"Lexical Move\\">\\"message\\": \\"Validation Failed\\"</span></span><br/>
}<br/>
</code></pre>
</body>
</html>"
`;

exports[`annotateWithDiff should work with diff 1`] = `
"<html>
<body>
<style>
    body {
        background-color: black;
        color: white;
    }

    .Add {
        background-color: rgba(0,255,0,0.3);
    }

    .Delete {
        background-color: rgba(255,0,0,0.3);
    }

    .Move {
        background-color: rgba(0,0,255,0.3);
    }

    .string_left {
        width: 50%;
        float: left;
    }
</style>
<pre class=\\"string_left\\"><code>{<br/>
<span class=\\"Textual Delete\\">  </span><span class=\\"Textual Move\\">  \\"message\\": \\"Validation Failed\\"</span><span class=\\"Textual Move\\"><span class=\\"Textual Move\\">,<br/>
</span></span><span class=\\"Textual Delete\\">  </span><span class=\\"Textual Move\\"><span class=\\"Textual Move\\">  \\"errors\\": [<br/>
    </span></span><span class=\\"Textual Delete\\">    </span><span class=\\"Textual Move\\"><span class=\\"Textual Move\\">{<br/>
</span></span><span class=\\"Textual Delete\\">      </span>      \\"resource\\": \\"Issue\\",<br/>
      <span class=\\"Textual Delete\\">      </span>\\"field\\": \\"title\\",<br/>
<span class=\\"Textual Delete\\">      </span>      \\"code\\": \\"missing_field\\"<br/>
    <span class=\\"Textual Delete\\">    </span>},<br/>
    <span class=\\"Textual Delete\\">    </span>{<br/>
<span class=\\"Textual Delete\\">      </span>      \\"resource\\": \\"Issue\\",<br/>
<span class=\\"Textual Delete\\">      </span>      \\"field\\": \\"milestone\\",<br/>
      <span class=\\"Textual Delete\\">      </span>\\"code\\": \\"invalid\\"<br/>
    <span class=\\"Textual Delete\\">    </span>}<br/>
  <span class=\\"Textual Delete\\">  ]</span><span class=\\"Textual Move\\"><br/>
}<br/>
</code></pre>
<pre class=\\"string_right\\"><code>{<br/>
<span class=\\"Textual Move\\"><span class=\\"Textual Move\\">  \\"errors\\": [<br/>
    </span></span><span class=\\"Textual Move\\"><span class=\\"Textual Move\\">{<br/>
</span></span>      \\"resource\\": \\"Issue\\",<br/>
      \\"field\\": \\"title\\",<br/>
      \\"code\\": \\"missing_field\\"<br/>
    },<br/>
    {<br/>
      \\"resource\\": \\"Issue\\",<br/>
      \\"field\\": \\"milestone\\",<br/>
      \\"code\\": \\"invalid\\"<br/>
    }<br/>
  <span class=\\"Textual Add\\">]</span><span class=\\"Textual Move\\"><span class=\\"Textual Move\\">,<br/>
</span></span><span class=\\"Textual Move\\">  \\"message\\": \\"Validation Failed\\"</span><span class=\\"Textual Move\\"><br/>
}<br/>
</code></pre>
</body>
</html>"
`;
